cli/command/container/run_test.go

```diff
@@ -217,6 +221,86 @@ func TestRunAttachTermination(t *testing.T) {
 	}
 }
 
+func TestRunPullTermination(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
```

Comment:
I generally prefer `defer cancel()`, for a couple of reasons:
- it's what I already expect to see looking at Go code, so less cognitive overhead
- together with the previous point, I'm more used to seeing `t.Cleanup` used for other things, such as removing files created for testing, or anything more complex.
- it's less cryptic (I know what to expect of a `defer`, but I have no idea what `t.Cleanup` will do/when it will do it.

Reply:
The behavior of `t.Cleanup` and `defer` in this context are the same.
- Will run in LIFO
- Will run on test panic

The difference comes in using `t.Cleanup` inside a `helper` function to clean something up only after the test has completed versus `defer` running after the `helper` function has completed 

For example
```go

func helper(t *testing.T) {
  t.Cleanup(func() {
    // something
  })
}

func TestSomething(t *testing.T) {
 helper(t)
 // ...
 return
 // cleanup something runs here
} 
```

I've been using `t.Cleanup` mostly, but it doesn't really matter unless you have the special case shown above.

Reply:
It is noteworthy that, with parallel subtests, the behavior differs quite a bit, since `t.Cleanup` waits for the tests to be done before running.

```go
// Cleanup registers a function to be called when the test (or subtest) and all its
// subtests complete. Cleanup functions will be called in last added,
// first called order.
```

Which means that with:
```go
func TestMeow(t *testing.T) {
	defer func() {
		println("bork")
	}()
	t.Cleanup(func() {
		println("meow")
	})

	t.Run("a test", func(t *testing.T) {
		t.Parallel()
		println("testing things!")
	})
}
```

We get:
```console
=== RUN   TestMeow
=== RUN   TestMeow/a_test
=== PAUSE TestMeow/a_test
bork
=== CONT  TestMeow/a_test
testing things!
meow
--- PASS: TestMeow (0.00s)
    --- PASS: TestMeow/a_test (0.00s)
PASS
```

It's not the case here that we have parallel subtests, so it's fine either way, but good to keep in mind.

Reply:
@Benehiko are you ok changing this one to a `defer` ? 

Laura's comment was one of the situations I had in mind, and I recall I was bit by that situation (`t.Parallel` was added somewhere and all of a sudden a test became flaky);

> It is noteworthy that, with parallel subtests, the behavior differs quite a bit, since t.Cleanup waits for the tests to be done before running.

Reply:
Sorry if I'm misunderstanding you, but wouldn't we want the cleanup to happen after all sub-tests completed (using t.Cleanup)? Doing a `defer` on a context cancel would cancel the sub-test (if it were marked as parallel).

Reply:
DOH! My bad I had my wires crossed completely, and read it in reverse (`defer` being the problematic one with `t.Parallel` and subtests, not `t.Cleanup`)

Yes, we should keep it as is then! 🙈 🙈 🙈